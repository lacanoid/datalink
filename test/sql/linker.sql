\pset null _null_

SET client_min_messages = notice;
SET search_path = public,datalink;

create table sample_datalinks6 (
  id serial,
  link datalink,
  link2 datalink
);

update datalink.columns
   set link_control='FILE', integrity='ALL',
       read_access='DB', write_access='TOKEN',
       recovery='YES', on_unlink='RESTORE'
 where table_name='sample_datalinks6' and column_name='link';

update datalink.columns
   set link_control='FILE', integrity='ALL'
 where table_name='sample_datalinks6' and column_name='link2';

insert into sample_datalinks6 (link)
values (dlvalue('/var/www/datalink/CHANGELOG.md','FS','Sample file datalink 1'));

update sample_datalinks6 set link2 = link;
update sample_datalinks6 set link = link2;
update sample_datalinks6 set link2 = null;

truncate sample_datalinks6;
insert into sample_datalinks6 (link)
values (dlvalue('/var/www/datalink/CHANGELOG.md','FS','Sample file datalink 2'));
insert into sample_datalinks6 (link)
values (dlvalue('http://www.debian.org/tmp/CHANGELOG.md',null,'Weblink'));

select regexp_replace(dlurlpath(link),'[a-z0-9\-]{10,}','xxx','g') as dlurlpath1
  from sample_datalinks6;
select dlurlpathonly(link) from sample_datalinks6;

update sample_datalinks6
   set link = dlnewcopy(link);

update sample_datalinks6 set link2 = link, link = null;
update sample_datalinks6 set link2 = null, link = link2;

create table sample_datalinks7 as 
select * 
  from sample_datalinks6;

truncate sample_datalinks6;


