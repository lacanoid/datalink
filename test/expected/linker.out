\pset null _null_
\set VERBOSITY default
\set CONTEXT never
SET client_min_messages = notice;
SET search_path = public,datalink;
create table sample_datalinks6 (
  id serial,
  link datalink,
  link2 datalink
);
update datalink.columns
   set link_control='FILE', integrity='ALL',
       read_access='DB', write_access='TOKEN',
       recovery='YES', on_unlink='RESTORE'
 where table_name='sample_datalinks6' and column_name='link';
NOTICE:  DATALINK DDL:TRIGGER on sample_datalinks6
update datalink.columns
   set link_control='FILE', integrity='ALL'
 where table_name='sample_datalinks6' and column_name='link2';
insert into sample_datalinks6 (link)
values (dlvalue('/var/www/datalink/CHANGELOG.md','FS','Sample file datalink 1'));
WARNING:  DATALINK WARNING - datalinker required
HINT:  make sure pg_datalinker process is running to finalize your commits
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
update sample_datalinks6 set link2 = link;
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
ERROR:  DATALINK EXCEPTION - external file already linked
DETAIL:  from sample_datalinks6.link as '/var/www/datalink/CHANGELOG.md'
CONTEXT:  PL/pgSQL function dl_file_link(file_path,dl_token,dl_lco,regclass,name) line 56 at RAISE
SQL statement "SELECT datalink.dl_file_link(dlurlpathonly(link),(link->>'b')::datalink.dl_token,link_options,regclass,column_name)"
PL/pgSQL function dl_datalink_ref(datalink,dl_lco,regclass,name) line 66 at PERFORM
PL/pgSQL function dl_trigger_table() line 73 at assignment
update sample_datalinks6 set link = link2;
NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
update sample_datalinks6 set link2 = null;
truncate sample_datalinks6;
insert into sample_datalinks6 (link)
values (dlvalue('/var/www/datalink/CHANGELOG.md','FS','Sample file datalink 2'));
WARNING:  DATALINK WARNING - datalinker required
HINT:  make sure pg_datalinker process is running to finalize your commits
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
insert into sample_datalinks6 (link)
values (dlvalue('http://www.debian.org/tmp/CHANGELOG.md',null,'Weblink'));
ERROR:  DATALINK EXCEPTION - invalid datalink construction
DETAIL:  INTEGRITY ALL can only be used with file URLs
HINT:  make sure you are using a file: URL scheme
CONTEXT:  PL/pgSQL function dl_datalink_ref(datalink,dl_lco,regclass,name) line 17 at RAISE
PL/pgSQL function dl_trigger_table() line 73 at assignment
select regexp_replace(dlurlpath(link),'[a-z0-9\-]{10,}','xxxx','g') as dlurlpath1
  from sample_datalinks6;
             dlurlpath1              
-------------------------------------
 /var/www/datalink/xxxx;CHANGELOG.md
(1 row)

select dlurlpathonly(link) from sample_datalinks6;
         dlurlpathonly          
--------------------------------
 /var/www/datalink/CHANGELOG.md
(1 row)

update sample_datalinks6
   set link = dlnewcopy(link);
NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
WARNING:  DATALINK WARNING - datalinker required
HINT:  make sure pg_datalinker process is running to finalize your commits
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
update sample_datalinks6 set link2 = link, link = null;
NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
update sample_datalinks6 set link2 = null, link = link2;
NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
WARNING:  DATALINK WARNING - datalinker required
HINT:  make sure pg_datalinker process is running to finalize your commits
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
create table sample_datalinks7 as 
select * 
  from sample_datalinks6;
NOTICE:  DATALINK DDL:TRIGGER on sample_datalinks7
truncate sample_datalinks6;
NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
insert into sample_datalinks6 (link) values (dlvalue('/var/www/datalink/CHANGELOG.md'));
WARNING:  DATALINK WARNING - datalinker required
HINT:  make sure pg_datalinker process is running to finalize your commits
NOTICE:  DATALINK LINK:/var/www/datalink/CHANGELOG.md
select link-'b' from sample_datalinks6; truncate sample_datalinks6;
                        ?column?                        
--------------------------------------------------------
 {"a": "file:///var/www/datalink/CHANGELOG.md", "r": 1}
(1 row)

NOTICE:  DATALINK UNLINK:/var/www/datalink/CHANGELOG.md
insert into my_table2
select dlvalue(filename)
  from sample_files
 where filename like '%.txt'
   and not (filename like '%X%' or filename like '%3.txt');
NOTICE:  DATALINK LINK:/var/www/datalink/test1.txt
NOTICE:  DATALINK LINK:/var/www/datalink/test2.txt
truncate my_table2;
NOTICE:  DATALINK UNLINK:/var/www/datalink/test1.txt
NOTICE:  DATALINK UNLINK:/var/www/datalink/test2.txt
-- test for foreign servers
select * from datalink.curl_get('file://zala/etc/issue');
ERROR:  Extension dblink is required for files on foreign servers.
CONTEXT:  PL/Perl function "curl_get"
create extension dblink;
select * from datalink.curl_get('file://zala/etc/issue');
ERROR:  Foreign server zala does not exist.
CONTEXT:  PL/Perl function "curl_get"
create extension postgres_fdw;
create server zala foreign data wrapper postgres_fdw options (dbname 'contrib_regression');
select * from datalink.curl_get('file://zala/etc/issue');
ERROR:  user mapping not found for "root" at line 38.
CONTEXT:  PL/Perl function "curl_get"
select * from datalink.curl_get('file://tiha/etc/issue');
ERROR:  Foreign server tiha does not exist.
CONTEXT:  PL/Perl function "curl_get"
create user mapping for current_user server zala;
select url,ok,rc,error from datalink.curl_get('file://zala/etc/issue');
          url          | ok | rc | error  
-----------------------+----+----+--------
 file://zala/etc/issue | t  |  0 | _null_
(1 row)

select url,ok,rc,error from datalink.curl_get('file://zala/etc/issueXXXXX');
            url             | ok | rc |            error             
----------------------------+----+----+------------------------------
 file://zala/etc/issueXXXXX | f  | 37 | Couldn't read a file:// file
(1 row)

-- test directories
update datalink.directory
   set dirname='www', 
       dirowner=current_role::regrole,
       dirurl='http://localhost/datalink/'
 where dirpath='/var/www/datalink/';
NOTICE:  DATALINK LINK:/var/www/datalink/
select dlvalue('hello.txt','www');
                         dlvalue                         
---------------------------------------------------------
 {"a": "file:///var/www/datalink/hello.txt", "t": "www"}
(1 row)

select dlvalue('','www');
                    dlvalue                     
------------------------------------------------
 {"a": "file:///var/www/datalink/", "t": "www"}
(1 row)

select dlvalue(NULL,'www');
                    dlvalue                     
------------------------------------------------
 {"a": "file:///var/www/datalink/", "t": "www"}
(1 row)

select dlvalue('http://localhost/datalink/CHANGELOG.md');
                        dlvalue                         
--------------------------------------------------------
 {"a": "file:///var/www/datalink/CHANGELOG.md", "m": 1}
(1 row)

select dlurlcomplete(dlvalue('http://localhost/datalink/CHANGELOG.md'));
             dlurlcomplete              
----------------------------------------
 http://localhost/datalink/CHANGELOG.md
(1 row)

select dlurlcompleteonly(dlvalue('http://localhost/datalink/CHANGELOG.md'));
           dlurlcompleteonly            
----------------------------------------
 http://localhost/datalink/CHANGELOG.md
(1 row)

select dlvalue('http://localhost/datalinkxxx/CHANGELOG.md');
                      dlvalue                       
----------------------------------------------------
 {"a": "http://localhost/datalinkxxx/CHANGELOG.md"}
(1 row)

select dlurlcomplete(dlvalue('http://localhost/datalinkxxx/CHANGELOG.md'));
               dlurlcomplete               
-------------------------------------------
 http://localhost/datalinkxxx/CHANGELOG.md
(1 row)

select dlurlcompleteonly(dlvalue('http://localhost/datalinkxxx/CHANGELOG.md'));
             dlurlcompleteonly             
-------------------------------------------
 http://localhost/datalinkxxx/CHANGELOG.md
(1 row)

insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink/','PUBLIC','SELECT');
insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink/','PUBLIC','REFERENCES');
begin;
create role dl_access_test;
select dirpath,privilege_type,grantee from datalink.access;
      dirpath       | privilege_type | grantee 
--------------------+----------------+---------
 /var/www/datalink/ | SELECT         | PUBLIC
 /var/www/datalink/ | REFERENCES     | PUBLIC
(2 rows)

insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink/','dl_access_test','UPDATE');
insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink/','PUBLIC','REFERENCES');
insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink/','dl_access_test','SELECT');
select dirpath,privilege_type,grantee from datalink.access;
      dirpath       | privilege_type |    grantee     
--------------------+----------------+----------------
 /var/www/datalink/ | SELECT         | PUBLIC
 /var/www/datalink/ | REFERENCES     | PUBLIC
 /var/www/datalink/ | UPDATE         | dl_access_test
 /var/www/datalink/ | SELECT         | dl_access_test
(4 rows)

delete from datalink.access where grantee = 'dl_access_test';
select dirpath,privilege_type,grantee from datalink.access;
      dirpath       | privilege_type | grantee 
--------------------+----------------+---------
 /var/www/datalink/ | SELECT         | PUBLIC
 /var/www/datalink/ | REFERENCES     | PUBLIC
(2 rows)

delete from datalink.access where grantee = 'PUBLIC';
select dirpath,privilege_type,grantee from datalink.access;
 dirpath | privilege_type | grantee 
---------+----------------+---------
(0 rows)

insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink/','dl_access_test','DELETE');
select dirpath,privilege_type,grantee from datalink.access;
      dirpath       | privilege_type |    grantee     
--------------------+----------------+----------------
 /var/www/datalink/ | DELETE         | dl_access_test
(1 row)

drop role dl_access_test;
rollback;
insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink/','PUBLIC','KIKI');
ERROR:  unrecognized privilege type: "KIKI"
CONTEXT:  PL/pgSQL function dl_trigger_access() line 26 at assignment
insert into datalink.access (dirpath,grantee,privilege_type) values ('/var/www/datalink_xxx/','PUBLIC','REFERENCES');
ERROR:  DATALINK EXCEPTION - directory not found
PATH:  /var/www/datalink_xxx/
DETAIL:  directory not found while modifying datalink.access
HINT:  add appropriate entry in table datalink.directory
CONTEXT:  PL/pgSQL function dl_trigger_access() line 14 at RAISE
select dirpath,privilege_type,grantee from datalink.access;
      dirpath       | privilege_type | grantee 
--------------------+----------------+---------
 /var/www/datalink/ | SELECT         | PUBLIC
 /var/www/datalink/ | REFERENCES     | PUBLIC
(2 rows)

